src/hyp_client/v1.py

```diff
@@ -8,44 +8,55 @@ def __init__(self, access_token, session=requests.Session()):
         self.session = session
         self.logger = logging.getLogger("hyp_python_client")
 
-    def try_assignment(self, participant_id, experiment_id, fallback):
-        response = self.assignment(participant_id, experiment_id)
-
-        if response["message"] == "success":
-            self.logger.info(f'Successfully got assignment for participant {participant_id} in experiment {experiment_id}.')
-            return response["payload"]["variant_name"]
-        else:
-            self.logger.warning(f'Failed to get assignment for participant {participant_id} in experiment {experiment_id}. Returning fallback {fallback}.')
-            return fallback
-
     def assignment(self, participant_id, experiment_id):
         response = self.session.post(
             f'https://app.onhyp.com/api/v1/assign/{participant_id}/{experiment_id}',
-            headers={'X_HYP_TOKEN': self.access_token},
+            headers={'X_HYP_TOKEN': self.access_token, 'Content-Type': 'application/json'},
         )
 
         result = response.json()
         result["status_code"] = response.status_code
 
         return result
 
-    def try_conversion(self, participant_id, experiment_id):
-        response = self.conversion(participant_id, experiment_id)
-
-        if response["message"] == "success":
-            self.logger.info(f'Successfully converted participant {participant_id} in experiment {experiment_id}.')
-            return response["payload"]["converted"]
-        else:
-            self.logger.warning(f'Failed to convert participant {participant_id} in experiment {experiment_id}. Returning False.')
-            return False
-
     def conversion(self, participant_id, experiment_id):
         response = self.session.patch(
             f'https://app.onhyp.com/api/v1/convert/{participant_id}/{experiment_id}',
-            headers={'X_HYP_TOKEN': self.access_token},
+            headers={'X_HYP_TOKEN': self.access_token, 'Content-Type': 'application/json'},
         )
 
         result = response.json()
         result["status_code"] = response.status_code
 
         return result
+
+    def try_api_call(self, method_name, participant_id, experiment_id, fallback):
+        if participant_id is None or experiment_id is None:
+            missing_data = []
+            if participant_id is None:
+                missing_data.append("participant ID")
+
+            if experiment_id is None:
+                missing_data.append("experiment ID")
+
+            missing_data_message = " and ".join(missing_data)
+            self.logger.warning(f'{method_name} failed due to missing {missing_data_message}. Returning fallback {fallback}.')
+            return fallback
+
+        response = getattr(self, method_name)(participant_id, experiment_id)
+
+        if response["message"] == "success":
+            self.logger.info(f'{method_name} successful for participant {participant_id} in experiment {experiment_id}.')
+            if method_name == "assignment":
```

Comment:
This `if` statement isn't extensible. If you were to add another API method with a `try_` method, you'd have to add two more lines of code to this `if` statement to handle the new method.

You could pass the field name as a parameter and use that to extract the data from the response. That way you eliminate the use of the `method_name`.

e.g.:
`def try_api_call(self, method_name, participant_id, experiment_id, response_field, fallback):`
...
`return response["payload"][field_name]`

Reply:
I generally tend to not optimize for extensibility right away. If we continue to add conditionals to this code it'll be signal that there is a more general abstraction that should be leaned into. Right now it's not clear that there will be more API calls to support and what the mechanics of those API calls will be. If that does turn out to be the case then I'll be more inclined to make this more dynamic.

There is certainly a lot of subjectivity here, I'm not saying that it would definitely be wrong to handle this as you suggest, just that I don't think the current imperative code is problematic.

Eg it's not as if this conditional is very large or complex, resulting in difficult to understand code.